File: src/main/java/com/linebot/service/tobuy/TobuyService.java
package com.linebot.service.tobuy;

import com.linebot.entity.Tobuy;
import com.linebot.repository.TobuyRepository;
import com.linebot.util.Utils;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Slf4j
@AllArgsConstructor
@Service
public class TobuyService {
    private TobuyRepository tobuyRepository;

    @Transactional(readOnly = true)
    public List<Tobuy> findByIsCompleted(String userId, String isCompleted) {
        return tobuyRepository.findByUserIdAndIsCompleted(userId, isCompleted);
    }

    @Transactional(readOnly = true)
    public Tobuy findByIdAndGoods(Integer id, String userId, String goods) {
        return tobuyRepository.findByIdAndUserIdAndGoodsAndIsCompleted(id, userId, goods, "0");
    }

    @Transactional
    public int insertByGoods(String userId, String goods) {
        Tobuy t = new Tobuy();
        t.setGoods(goods);
        t.setIsCompleted("0");
        t.setCreatedDate(Utils.now());
        t.setUpdatedDate(Utils.now());
        t.setUserId(userId);
        tobuyRepository.save(t);
        return 1;
    }

    @Transactional
    public int updateCompleted(Tobuy tobuy) {
        tobuy.setIsCompleted("1");
        tobuy.setUpdatedDate(Utils.now());
        tobuyRepository.save(tobuy);
        return 1;
    }

    @Transactional
    public int updateCompleted(List<Tobuy> tobuys) {
        return tobuys.stream().mapToInt(tobuy -> {
            tobuy.setIsCompleted("1");
            tobuy.setUpdatedDate(Utils.now());
            tobuyRepository.save(tobuy);
            return 1;
        }).sum();
    }
}

File: src/main/kotlin/com/linebot/service/tobuy/TobuyService.kt
package com.linebot.service.tobuy

import com.linebot.entity.Tobuy
import com.linebot.repository.TobuyRepository
import com.linebot.util.Utils
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class TobuyService(
    private val tobuyRepository: TobuyRepository
) {

    @Transactional(readOnly = true)
    fun findByIsCompleted(userId: String, isCompleted: String): List<Tobuy> {
        return tobuyRepository.findByUserIdAndIsCompleted(userId, isCompleted)
    }

    @Transactional(readOnly = true)
    fun findByIdAndGoods(id: Int?, userId: String, goods: String): Tobuy? {
        return tobuyRepository.findByIdAndUserIdAndGoodsAndIsCompleted(id, userId, goods, "0")
    }

    @Transactional
    fun insertByGoods(userId: String, goods: String): Int {
        tobuyRepository.save(Tobuy(
                goods = goods,
                isCompleted = "0",
                createdDate = Utils.now(),
                updatedDate = Utils.now(),
                userId = userId
        ))
        return 1
    }

    @Transactional
    fun updateCompleted(tobuy: Tobuy): Int {
        tobuy.isCompleted = "1"
        tobuy.updatedDate = Utils.now()
        tobuyRepository.save(tobuy)
        return 1
    }

    @Transactional
    fun updateCompleted(tobuys: List<Tobuy>): Int {
        return tobuys.map {
            it.isCompleted = "1"
            it.updatedDate = Utils.now()
            tobuyRepository.save(it)
        }.count()
    }
}
